### PR TITLE
aSync annotators + jobMetadata + progressBar

### PR DESCRIPTION
**Pull request by bartcharbon**
- add progress bar component
- add jobMetadata
- progress bar for annotators
- quartz based asynchronous annotators
- updated annotator ui to cope with async
- move copy function outside of annotator framework

--Pull request can be tested and reviewed, but we might want to keep it open until a complementing selenium test pull request is available. --
https://github.com/molgenis/molgenis-selenium/pull/50 selenium request is ready.

### CODE REVIEW

**Comment by tommydeboer** [comment]
test this please

**Comment by tommydeboer** [comment]
test this please thank you

File: molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/AnnotationJob.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.List;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class AnnotationJob implements org.quartz.Job
```
Comment (tommydeboer) [inline molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/AnnotationJob.java]: Why not import ```org.quartz.Job``` ?

File: molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/utils/AnnotatorUtils.java [hunk 2]
```diff
@@ -10,9 +10,6 @@
 
 import autovalue.shaded.com.google.common.common.collect.Iterables;
 
-/**
- * Created by charbonb on 11/03/15.
- */
```
Comment (tommydeboer) [inline molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/utils/AnnotatorUtils.java]: Removing evidence, are we?

File: molgenis-dataexplorer/src/main/resources/templates/view-dataexplorer-mod-annotators.ftl [hunk 3]
```diff
@@ -1,4 +1,54 @@
...
+	                                var container = $('#annotateRun');
+	
+	                                if(entry.status!=="RUNNING"){
+	             	                     window.location.replace("?entity=${entityName}");
+	                                }
+	                                else{
+	                                	if(ProgressBar && ProgressBar.isMounted()) {
+                                            var progress = ((entry.progressInt/entry.progressMax)*100)
+                                            console.log(entry.progressMessage);
+                                            console.log(progress);
```
Comment (tommydeboer) [inline molgenis-dataexplorer/src/main/resources/templates/view-dataexplorer-mod-annotators.ftl]: remove debug logging

**Comment by tommydeboer** [comment]
There's still an annotator running when the progress bar is at 100%

File: molgenis-data/src/main/java/org/molgenis/data/jobs/JobMetaDataMetaData.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+				.setNillable(false);
+		addAttribute(JobMetaData.START_DATE).setDataType(MolgenisFieldTypes.DATETIME).setLabel("Job start date")
+				.setNillable(true);
+		addAttribute(JobMetaData.END_DATE).setDataType(MolgenisFieldTypes.DATETIME).setLabel("Job end date").setNillable(true);
+		addAttribute(JobMetaData.PROGRESS_INT).setDataType(MolgenisFieldTypes.INT).setLabel("Number of entities processed")
+				.setNillable(true);
+		addAttribute(JobMetaData.PROGRESS_MESSAGE).setDataType(MolgenisFieldTypes.TEXT).setLabel("Progress message")
+				.setNillable(true);
+		addAttribute(JobMetaData.PROGRESS_MAX).setDataType(MolgenisFieldTypes.INT)
+				.setLabel("Maximum number of entities to be processed").setNillable(true);
```
Comment (tommydeboer) [inline molgenis-data/src/main/java/org/molgenis/data/jobs/JobMetaDataMetaData.java]: After some discussion: put the ```PROGRESS_INT``` and ```PROGRESS_MAX``` attributes next to each other and change the descriptions to something like "progress" and "maximum progress".

**Comment by tommydeboer** [comment]
Should we add a story to make it possible to view abstract entities in the explorer? Would come in handy if you want to look at all Jobs, irrespective of type.

File: molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/AnnotationJob.java (hunk 1 above)
Comment (bartcharbon) [inline molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/AnnotationJob.java]: This was bacause JobMetaData was also called Job at the time... I'll update

File: molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/utils/AnnotatorUtils.java (hunk 2 above)
Comment (bartcharbon) [inline molgenis-data-annotators/src/main/java/org/molgenis/data/annotation/utils/AnnotatorUtils.java]: unfortunately git will for always remember

**Comment by bartcharbon** [comment]
Story still sound like a good idea, however there will be a dashboard for the jobs I think, so no need to do this for that use case.

**Pull request merged**